File: term 2/Segment tree/B.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class B2 {

    static class SegmentTree {
        int st[];

        SegmentTree(int a[], int n) {
            st = new int[4 * n];
            buildST(a, 0, n - 1, 0);
        }

        int getMiddle(int l, int r) {
            return l + (r - l) / 2;
        }

        int getSum(int ss, int se, int qs, int qe, int si) {

            if (qs <= ss && qe >= se)
                return st[si];

            if (se < qs || ss > qe)
                return 0;

            int mid = getMiddle(ss, se);
            return getSum(ss, mid, qs, qe, 2 * si + 1) +
                    getSum(mid + 1, se, qs, qe, 2 * si + 2);
        }

        void updateValue(int ss, int se, int i, int diff, int si) {
            if (i < ss || i > se)
                return;
            st[si] = st[si] + diff;
            if (se != ss) {
                int mid = getMiddle(ss, se);
                updateValue(ss, mid, i, diff, 2 * si + 1);
                updateValue(mid + 1, se, i, diff, 2 * si + 2);
            }
        }

        void update(int arr[], int n, int i, int new_val) {
            if (i < 0 || i > n - 1) {
                return;
            }
            int diff = new_val - arr[i];
            arr[i] = new_val;
            updateValue(0, n - 1, i, diff, 0);
        }

        int get(int n, int qs, int qe) {
            if (qs < 0 || qe > n - 1 || qs > qe) {
                return -1;
            }
            return getSum(0, n - 1, qs, qe, 0);
        }

        int buildST(int arr[], int ss, int se, int si) {
            if (ss == se) {
                st[si] = arr[ss];
                return arr[ss];
            }
            int mid = getMiddle(ss, se);
            st[si] = buildST(arr, ss, mid, si * 2 + 1) +
                    buildST(arr, mid + 1, se, si * 2 + 2);
            return st[si];
        }
    }

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(new File("rsq.in"));
        FileWriter fileWriter = new FileWriter(new File("rsq.out"));

        int n = scanner.nextInt();
        int a[] = new int[n];
        for (int i = 0; i < n; i++) a[i] = scanner.nextInt();

        SegmentTree segmentTree = new SegmentTree(a, n);
        String s = "";
        while (scanner.hasNext()) {
            s = scanner.next();
            switch (s) {
                case "sum":
                    int l = scanner.nextInt() - 1;
                    int r = scanner.nextInt() - 1;
                    fileWriter.write(segmentTree.get(n, l, r) + "\n");
                    break;
                case "set":
                    int i = scanner.nextInt() - 1;
                    int x = scanner.nextInt();
                    segmentTree.update(a, n, i, x);
                    break;
            }
        }

        fileWriter.flush();
    }
}

File: term 2/Segment tree/B.cpp
#include <iostream>

using namespace std;

int getMiddle(int l, int r) {
    return l + (r - l) / 2;
}

void updating(long long* st, int tl, int tr, int i, long long diff, int ind) {
    if (i < tl || i > tr) {
        return;
    }

    st[ind] = st[ind] + diff;
    if (tr != tl) {
        int mid = getMiddle(tl, tr);
        updating(st, tl, mid, i, diff, 2 * ind + 1);
        updating(st, mid + 1, tr, i, diff, 2 * ind + 2);
    }
}

void update(long long* a, long long* st, int n, int i, long long new_Value) {
    long long diff = new_Value - a[i];
    a[i] = new_Value;
    updating(st, 0, n - 1, i, diff, 0);
}

long long buildTree(long long* a, int tl, int tr, long long* st, int ind) {
    if (tl == tr) {
        st[ind] = a[tl];
        return a[tl];
    }

    int mid = getMiddle(tl, tr);
    st[ind] = buildTree(a, tl, mid, st, ind * 2 + 1) + buildTree(a, mid + 1, tr, st, ind * 2 + 2);

    return st[ind];
}

long long sum(long long* st, int tl, int tr, int l, int r, int ind) {
    if (l <= tl && r >= tr) {
        return st[ind];
    }

    if (tr < l || tl > r) {
        return 0;
    }

    int mid = getMiddle(tl, tr);

    return sum(st, tl, mid, l, r, 2 * ind + 1) + sum(st, mid + 1, tr, l, r, 2 * ind + 2);

}

long long* buildingTree(long long* a, int n) {
    long long* st = new long long[4 * n];
    buildTree(a, 0, n - 1, st, 0);
    return st;
}

long long get(long long* st, int n, int l, int r) {
    return sum(st, 0, n - 1, l, r, 0);
}


int main() {
    freopen("rsq.in", "r", stdin);
    freopen("rsq.out", "w", stdout);

    int n;
    cin >> n;

    long long a[n];
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    long long* st = buildingTree(a, n);
    string s;

    while (cin >> s) {
        if (s == "sum") {
            int i, j;
            cin >> i;
            cin >> j;
            cout << get(st, n, i - 1, j - 1) << "\n";
        }
        if (s == "set") {
            int i;
            long long x;
            cin >> i;
            cin >> x;
            update(a, st, n, i - 1, x);
        }
    }

    return 0;
}